src/Umbraco.Web/PublishedContentQuery.cs

```diff
@@ -224,14 +225,36 @@ private IPublishedContent TypedDocumentById(int id, ContextualPublishedCache cac
         {
             var doc = cache.GetById(id);
             return doc;
-        }
+        }
+
+        private static ConcurrentDictionary<Guid, int> _guidToIntLoopkup;
 
         private IPublishedContent TypedDocumentById(Guid id, ContextualPublishedCache cache)
-        {
-            // todo: in v8, implement in a more efficient way
-            var legacyXml = UmbracoConfig.For.UmbracoSettings().Content.UseLegacyXmlSchema;
-            var xpath = legacyXml ? "//node [@key=$guid]" : "//* [@key=$guid]";
-            var doc = cache.GetSingleByXPath(xpath, new XPathVariable("guid", id.ToString()));
+        {
+            // We create the lookup when we first need it
+            if (_guidToIntLoopkup == null)
+                _guidToIntLoopkup = new ConcurrentDictionary<Guid, int>();
+
+            IPublishedContent doc;
+
+            // Check if the lookup contains the GUID/INT value
+            if (_guidToIntLoopkup.TryGetValue(id, out int nodeId) == false)
+            {
+                // If not, then we perform an inefficient XPath for the GUID
+
+                // todo: in v8, implement in a more efficient way
+                doc = cache.GetSingleByXPath("//*[@key=$guid]", new XPathVariable("guid", id.ToString()));
+
+                // When we have the node, we add the GUID/INT value to the lookup
+                if (doc != null)
+                    _guidToIntLoopkup.TryAdd(id, doc.Id);
```

Comment:
Since we know that the key/value pair here are correct. There's no need for `TryAdd` since you are not checking the result anyway. `_guidToIntLoopkup[id] = doc.Id` should be sufficient.